src/ast_editor.ml

```diff
@@ -0,0 +1,514 @@
+
+let reload_pp_doc instance ~document =
+  let open Promise.Syntax in
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let visibleTextEditors = Window.visibleTextEditors () in
+  let origin_uri =
+    match
+      Ast_editor_state.find_original_doc_by_pp_uri
+        ~uri_string:(doc_string_uri ~document)
+        (Ast_editor_state.get_origin_to_pp_doc_map ast_editor_state)
+    with
+    | Some x -> x
+    | None -> failwith "Failed finding the original document URI."
+  in
+  let* original_document =
+    Workspace.openTextDocument (`Uri (Uri.parse origin_uri ()))
+  in
+  match
+    List.find
+      ~f:(fun editor -> document_eq (TextEditor.document editor) document)
+      visibleTextEditors
+  with
+  | Some _ ->
+    let+ (_ : bool) =
+      Ast_editor_state.set_origin_changed ast_editor_state
+        ~key:(doc_string_uri ~document) ~data:false;
+      replace_document_content
+        ~content:
+          (Ppx_tools.get_reparsed_code_from_pp_file ~document:original_document)
+        ~document
+    in
+    0
+  | None -> Promise.resolve 1
+
+let manage_choice instance choice ~document : int Promise.t =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match Ppx_tools.Pp_path.project_root_path () with
+  | None ->
+    show_message `Error "%s" "Error : project root wasn't found";
+    Promise.resolve 1
+  | Some root -> (
+    let path = Path.of_string root in
+    let buildCmd () =
+      Cmd.run ~cwd:path (Cmd.Shell "eval $(opam env); dune build")
+    in
+    let rec build_project () =
+      let open Promise.Syntax in
+      let* res = buildCmd () in
+      if res.exitCode = 0 then
+        if
+          Map.existsi
+            (Ast_editor_state.get_pp_doc_to_changed_origin_map ast_editor_state)
+            ~f:(fun ~key ~data:_ -> String.equal key (doc_string_uri ~document))
+        then
+          reload_pp_doc instance ~document
+        else
+          open_pp_doc instance ~document
+      else
+        let* perror =
+          Window.showErrorMessage
+            ~message:"Building project failed, fix project errors and retry."
+            ~choices:[ ("Retry running `dune build`", 0); ("Abandon", 1) ]
+            ()
+        in
+        match perror with
+        | Some 0 -> build_project ()
+        | _ -> Promise.resolve 1
+    in
+    match choice with
+    | Some 0 -> build_project ()
+    | _ -> Promise.resolve 1)
+
+let manage_open_failure instance ~document =
+  let open Promise.Syntax in
+  let* choice =
+    Window.showInformationMessage
+      ~message:
+        ("Seems like the file '"
+        ^ Ppx_tools.Pp_path.relative_document_path ~document
+        ^ "' haven't been preprocessed yet.")
+      ~choices:[ ("Run `dune build`", 0); ("Abandon", 1) ]
+      ()
+  in
+  manage_choice instance choice ~document
+
+let open_preprocessed_doc_to_the_side instance ~document =
+  try open_pp_doc instance ~document with
+  | _ -> manage_open_failure instance ~document
```

Comment:
Again, we shouldn't be catching all exceptions.

Reply:
Changed to catching `Sys_error` when file is not found in b6f514d1400b2464909d6a0657899bf10b6cbcc5